src/html/Quoting.cc

```diff
@@ -9,81 +9,74 @@
+    const size_t maxEscapeLength = 7;
+    /* Encode control chars just to be on the safe side, and make
+     * sure all 8-bit characters are encoded to protect from buggy
+     * clients
      */
-    {
-        '<', "&lt;"
-    },
-    {
-        '>', "&gt;"
-    },
-    {
-        '"', "&quot;"
-    },
-    {
-        '&', "&amp;"
-    },
-    {
-        '\'', "&#39;"
-    },
-    {
-        0, NULL
+    for (uint32_t ch = 1; ch < 256; ++ch) {
+        if ((ch <= 0x1F || ch >= 0x7f) && ch != '\n' && ch != '\r' && ch != '\t') {
+            escapeMap[ch] = static_cast<char *>(xcalloc(maxEscapeLength, 1));
+            snprintf(const_cast<char*>(escapeMap[ch]), sizeof escapeMap[ch], "&#%d;", static_cast<int>(ch));
+        }
     }
-};
+    for (auto &pair: escapePairs) {
+        escapeMap[pair.first] = static_cast<char *>(xcalloc(maxEscapeLength, 1));
+        xstrncpy(const_cast<char*>(escapeMap[pair.first]), pair.second, maxEscapeLength);
+    }
+}
+HtmlSpecialCharacters::~HtmlSpecialCharacters()
+{
+    for (auto &escape: escapeMap) {
+        xfree(escape);
+    }
+}
 
-/*
- *  html_do_quote - Returns a static buffer containing the quoted
- *  string.
- */
 char *
 html_quote(const char *string)
 {
-    static char *buf = nullptr;
+    // for syntax, see https://html.spec.whatwg.org/#character-references
+    static const auto htmlSpecialCharacters = HtmlSpecialCharacters();
+
+    static char *buf;
     static size_t bufsize = 0;
     const char *src;
     char *dst;
-    int i;
 
-    /* XXX This really should be implemented using a MemPool, but
-     * MemPools are not yet available in lib...
+    /* XXX This really should be implemented using a MemPool
      */
-    if (!buf || strlen(string) * 6 > bufsize) {
+    if (buf == nullptr || strlen(string) * 6 > bufsize) {
         xfree(buf);
         bufsize = strlen(string) * 6 + 1;
-        buf = static_cast<char *>(xcalloc(bufsize, 1));
+        buf = static_cast<char*>(xcalloc(bufsize, 1));
     }
     for (src = string, dst = buf; *src; src++) {
-        const char *escape = NULL;
+        const char *escape = nullptr;
         const unsigned char ch = *src;
 
-        /* Walk thru the list of HTML Entities that must be quoted to
-         * display safely
-         */
-        for (i = 0; htmlstandardentities[i].code; i++) {
-            if (ch == htmlstandardentities[i].code) {
-                escape = htmlstandardentities[i].quote;
-                break;
-            }
-        }
-        /* Encode control chars just to be on the safe side, and make
-         * sure all 8-bit characters are encoded to protect from buggy
-         * clients
-         */
-        if (!escape && (ch <= 0x1F || ch >= 0x7f) && ch != '\n' && ch != '\r' && ch != '\t') {
-            static char dec_encoded[7];
-            snprintf(dec_encoded, sizeof dec_encoded, "&#%3d;", (int) ch);
-            escape = dec_encoded;
-        }
+        escape = htmlSpecialCharacters[ch];
+
         if (escape) {
```

Comment:
```suggestion
        if (const auto escape = htmlSpecialCharacters[ch]) {
```

... and drop the wrong-scope `escape` declaration above.